Add NotRouteConstraint for logical negation in route constraints

sami-daniel opened the pull request:
# Add NotRouteConstraint for logical negation in route constraints

Implements a new route constraint that provides logical negation capabilities
for existing route constraints. This enables developers to create routes that
match when specific constraints do NOT match.

Key Features:
- Single constraint negation: not(int) matches non-integers
- Multiple constraint negation: not(int;bool) matches neither integers nor booleans
- Nested negation support: not(not(int)) for double negation (equivalent to int)
- Parameterized constraints: not(min(18)) for values under 18 or non-numeric
- File extension negation: not(file) for values without file extensions
- String pattern negation: not(alpha) for non-alphabetic values
and other patterns

Examples:
- /users/{id:not(int)} - matches /users/john but not /users/123
- /files/{name:not(file)} - matches /files/readme but not /files/doc.txt
- /ages/{age:not(min(18))} - matches /ages/16 but not /ages/25
- /values/{val:not(int;bool)} - matches /values/text but not /values/123 or /values/true

Implementation:
- Added NotRouteConstraint class with XML documentation
- Supports both IRouteConstraint and IParameterLiteralNodeMatchingPolicy
- Added to RouteOptions default constraint map with 'not' token

Testing:
- Tests for single/multiple constraints, nested negation, edge cases
- Integration tests with HttpContext and service provider scenarios
- Validation of argument null exceptions and error handling
- Performance tests for literal matching optimization

This enables more flexible route patterns by allowing inverse constraint logic.

It can also be used to escape internal implementation details when
building route templates.

Fixes #62278

sami-daniel committed:
Change startvscode.sh file permissions to make it executable

captainsafia commented:
@sami-daniel These failures appear to be flakes in our build and not related to your change. I can kick the build here.

captainsafia commented on src/Http/Routing/src/Matching/DefaultEndpointSelector.cs:
```diff
@@ -60,6 +60,8 @@ private static void ProcessFinalCandidates(
         Endpoint? endpoint = null;
         RouteValueDictionary? values = null;
         int? foundScore = null;
+        var candidatesWithSameScore = new List<CandidateState>();
```
Can we initialize this the first time two candidates with the same score are detected?

captainsafia commented on src/Http/Routing/test/UnitTests/Matching/DfaMatcherTest.cs:
```diff
@@ -58,6 +58,67 @@ private DataSourceDependentMatcher CreateDfaMatcher(
+        // Assert 2
+        Assert.Same(endpoint2, httpContext.GetEndpoint());
+    }
+
```
Can we add a test for the case where the match is still ambiguous if there are specificity constraints that match?

captainsafia commented on src/Http/Routing/src/Matching/DefaultEndpointSelector.cs:
```diff
@@ -104,6 +119,104 @@ private static void ProcessFinalCandidates(
+        {
+            // Strong typed constraints that are very restrictive and has
+            // the highest specificity
+            "guid" => 100,
```
Using .Equals with a StringComparison to ignore casing might be more consistent than calling ToLowerInvariant for every constraint.

captainsafia reviewed (commented):
Thanks for taking the time to work on this change, @sami-daniel!

I left some notes on the current implementation but have two big picture thoughts:

- Given the nature of the change, we'll likely have to document this as a breaking behavioral change in routing given it changes the contract the router makes.
- I wonder if it might be better to support this feature by changing the logic that calculates precedence for route templates to account for constraints ([ref](https://github.com/dotnet/aspnetcore/blob/main/src/Http/Routing/src/Template/RoutePrecedence.cs)).

Cc: @javiercn @JamesNK for additional routing expertise

javiercn commented:
> Thanks for taking the time to work on this change, @sami-daniel!
> 
> I left some notes on the current implementation but have two big picture thoughts:
> 
> * Given the nature of the change, we'll likely have to document this as a breaking behavioral change in routing given it changes the contract the router makes.
> * I wonder if it might be better to support this feature by changing the logic that calculates precedence for route templates to account for constraints ([ref](https://github.com/dotnet/aspnetcore/blob/main/src/Http/Routing/src/Template/RoutePrecedence.cs)).
> 
> Cc: @javiercn @JamesNK for additional routing expertise

Few things.

I don't believe that the endpoint selector is the right place to make this change. Factoring it on the route precedence would make more sense.

That said, I think this change is problematic for several reasons:
* It impacts how routing behaves, which is not an ASP.NET Core only concern (precedence needs to be accounted for in the Blazor router).
* Constraints allow general code execution to participate in the matching process, ergo it's really hard if not impossible to correctly set a precedence, especially when custom user code is involved.
* Finally, I don't see general applicability for having to distinguish between two routes based uniquely on the constraints. It feels that the change adds significant overhead and complexity just to cover a "niche" scenario.
* In many cases, you can disambiguate in different ways by negating one of the constraints in one route. That to me feels like a more solid approach that keeps the routing behavior deterministic and doesn't involve having to "guess" the priorities that constraints impose.
* If we would want to take it further, we could introduce a pseudo-constraint to negate any existing constraint applied to a route like `:not(:int)` to address these scenarios.

I think before we even consider a change in this area, we need to discuss whether this is even a change that we want to make at all, and that should be something we discuss on an issue rather than on a PR.

sami-daniel commented:
> This is a hot path. Are we adding allocations?
> 
> Why do we need this logic in the default selector instead of one you can opt into?

Yes, we are adding allocations, and worse. We are copying the candidateState unnecessarily to the list when we could use it by reference. I will change that.

sami-daniel commented:
> * I wonder if it might be better to support this feature by changing the logic that calculates precedence for route templates to account for constraints ([ref](https://github.com/dotnet/aspnetcore/blob/main/src/Http/Routing/src/Template/RoutePrecedence.cs)).



> Thanks for taking the time to work on this change, @sami-daniel!
> 
> I left some notes on the current implementation but have two big picture thoughts:
> 
> * Given the nature of the change, we'll likely have to document this as a breaking behavioral change in routing given it changes the contract the router makes.
> * I wonder if it might be better to support this feature by changing the logic that calculates precedence for route templates to account for constraints ([ref](https://github.com/dotnet/aspnetcore/blob/main/src/Http/Routing/src/Template/RoutePrecedence.cs)).
> 
> Cc: @javiercn @JamesNK for additional routing expertise

Placing these changes in the [DefaultEndpointSelector](https://source.dot.net/#Microsoft.AspNetCore.Routing/Matching/[DefaultEndpointSelector.cs](https://source.dot.net/#Microsoft.AspNetCore.Routing/Matching/DefaultEndpointSelector.cs))  was a mistake, as it's not actually the source of the problem. The issue lies with its caller — the [DFAMatcher](https://source.dot.net/#Microsoft.AspNetCore.Routing/Matching/DfaMatcher.cs). The DefaultEndpointSelector is just the default implementation used by the engine when no custom endpoint selector is defined. In contrast, the DFAMatcher cannot be replaced or customized with its own selector, so it makes more sense for the changes to be applied there. The real problem is that the DFAMatcher does not consider the weight of constraints, as described on the pr — it only checks whether they are valid. 

I confess that I did not notice or pay attention to the use of [RoutePrecedence](https://source.dot.net/#Microsoft.AspNetCore.Routing/Template/RoutePrecedence.cs) in the code. From what I saw, it is used in the [DFAMatcher builder](https://source.dot.net/#Microsoft.AspNetCore.Routing/Matching/[DfaMatcherBuilder.cs](https://source.dot.net/#Microsoft.AspNetCore.Routing/Matching/DfaMatcherBuilder.cs,541),541), and in some [other places](https://source.dot.net/#Microsoft.AspNetCore.Routing/Template/[RoutePrecedence](https://source.dot.net/#Microsoft.AspNetCore.Routing/Template/RoutePrecedence.cs,62698bc323c9e1c9,references).cs,62698bc323c9e1c9,references) that I am not sure are related. I am not sure if it makes sense to make changes to RoutePrecedence, since it is not the direct cause of the problem. I would need to study a bit more to see where it applies...

sami-daniel commented:
> > Thanks for taking the time to work on this change, @sami-daniel!
> > I left some notes on the current implementation but have two big picture thoughts:
> > 
> > * Given the nature of the change, we'll likely have to document this as a breaking behavioral change in routing given it changes the contract the router makes.
> > * I wonder if it might be better to support this feature by changing the logic that calculates precedence for route templates to account for constraints ([ref](https://github.com/dotnet/aspnetcore/blob/main/src/Http/Routing/src/Template/RoutePrecedence.cs)).
> > 
> > Cc: @javiercn @JamesNK for additional routing expertise
> 
> Few things.
> 
> I don't believe that the endpoint selector is the right place to make this change. Factoring it on the route precedence would make more sense.
> 
> That said, I think this change is problematic for several reasons:
> 
> * It impacts how routing behaves, which is not an ASP.NET Core only concern (precedence needs to be accounted for in the Blazor router).
> * Constraints allow general code execution to participate in the matching process, ergo it's really hard if not impossible to correctly set a precedence, especially when custom user code is involved.
> * Finally, I don't see general applicability for having to distinguish between two routes based uniquely on the constraints. It feels that the change adds significant overhead and complexity just to cover a "niche" scenario.
> * In many cases, you can disambiguate in different ways by negating one of the constraints in one route. That to me feels like a more solid approach that keeps the routing behavior deterministic and doesn't involve having to "guess" the priorities that constraints impose.
> * If we would want to take it further, we could introduce a pseudo-constraint to negate any existing constraint applied to a route like `:not(:int)` to address these scenarios.
> 
> I think before we even consider a change in this area, we need to discuss whether this is even a change that we want to make at all, and that should be something we discuss on an issue rather than on a PR.

https://github.com/dotnet/aspnetcore/pull/62756#issuecomment-3089107701

sami-daniel commented:
As @javiercn said, perhaps it would be better, instead of adding overhead to `DFAMatcher` or the endpoint selector itself with a very niche change with little gain and a lot of tradeoff, to abstract this and leave control to the user. A new constraint that negates (`:not(:int))`, `:not(:guid))`) other constraints could be interesting to increase the number of possible templates and provide more flexibility with minimal cost. From another point of view, perhaps an exhaustive regex could handle a `:not(:guid))`, but I think the intention is to make things easier and not more complicated. I think it's worth pursuing this path. It seems like a complex challenge, but interesting to consider. I would like to work on it. If you have any ideas 😎

sami-daniel committed:
Add NotRouteConstraint implementation and associated tests

Implements a new route constraint that provides logical negation capabilities
for existing route constraints. This enables developers to create routes that
match when specific constraints do NOT match.

Key Features:
- Single constraint negation: not(int) matches non-integers
- Multiple constraint negation: not(int;bool) matches neither integers nor booleans
- Nested negation support: not(not(int)) for double negation (equivalent to int)
- Parameterized constraints: not(min(18)) for values under 18 or non-numeric
- File extension negation: not(file) for values without file extensions
- String pattern negation: not(alpha) for non-alphabetic values
and other patterns

Examples:
- /users/{id:not(int)} - matches /users/john but not /users/123
- /files/{name:not(file)} - matches /files/readme but not /files/doc.txt
- /ages/{age:not(min(18))} - matches /ages/16 but not /ages/25
- /values/{val:not(int;bool)} - matches /values/text but not /values/123 or /values/true

Implementation:
- Added NotRouteConstraint class with XML documentation
- Supports both IRouteConstraint and IParameterLiteralNodeMatchingPolicy
- Added to RouteOptions default constraint map with 'not' token

Testing:
- Tests for single/multiple constraints, nested negation, edge cases
- Integration tests with HttpContext and service provider scenarios
- Validation of argument null exceptions and error handling
- Performance tests for literal matching optimization

This enables more flexible route patterns by allowing inverse constraint logic.

It can also be used to escape internal implementation details when
building route templates.

sami-daniel commented on src/Http/Routing/src/Constraints/NotRouteConstraint.cs:
```diff
@@ -0,0 +1,297 @@
+        // FIXME: I'm not sure if this is a good thing to do because
+        // it requires weak spreading between the ConstraintMap and
+        // RouteOptions. It doesn't seem appropriate to create two
+        // identical variables for this...
```
I agree Mr. copilot 👍

sami-daniel commented:
> > As @javiercn said, perhaps it would be better, instead of adding overhead to `DFAMatcher` or the endpoint selector itself with a very niche change with little gain and a lot of tradeoff, to abstract this and leave control to the user. A new constraint that negates (`:not(:int))`, `:not(:guid))`) other constraints could be interesting to increase the number of possible templates and provide more flexibility with minimal cost. From another point of view, perhaps an exhaustive regex could handle a `:not(:guid))`, but I think the intention is to make things easier and not more complicated. I think it's worth pursuing this path. It seems like a complex challenge, but interesting to consider. I would like to work on it. If you have any ideas 😎
> 
> The other point about this is that there is already a mechanism to disambiguate, which is `Order`. I don't think we want any sort of mechanism with constraint "priorities", that's what Order is for. If we accept that, we might choose to use an analyzer instead to detect potentially ambiguous routes and suggest applying order. I think that's a reasonable compromise and doesn't involve changing the runtime. It won't cover 100% of the cases, but It'll likely cover the vast majority of them.



> > As @javiercn said, perhaps it would be better, instead of adding overhead to `DFAMatcher` or the endpoint selector itself with a very niche change with little gain and a lot of tradeoff, to abstract this and leave control to the user. A new constraint that negates (`:not(:int))`, `:not(:guid))`) other constraints could be interesting to increase the number of possible templates and provide more flexibility with minimal cost. From another point of view, perhaps an exhaustive regex could handle a `:not(:guid))`, but I think the intention is to make things easier and not more complicated. I think it's worth pursuing this path. It seems like a complex challenge, but interesting to consider. I would like to work on it. If you have any ideas 😎
> 
> The other point about this is that there is already a mechanism to disambiguate, which is `Order`. I don't think we want any sort of mechanism with constraint "priorities", that's what Order is for. If we accept that, we might choose to use an analyzer instead to detect potentially ambiguous routes and suggest applying order. I think that's a reasonable compromise and doesn't involve changing the runtime. It won't cover 100% of the cases, but It'll likely cover the vast majority of them.

It makes sense. I talked a bit about this in [referenced issue](https://github.com/dotnet/aspnetcore/issues/62278#issuecomment-3197599472), and it does. I'm not familiar with `Order`, but I found something very similar, which is a route precedence mechanism. It basically does that, haha. It calculates the precedence of routes selected by the DFA Build or something like that, so it would be reinventing the wheel. I ended up creating `NotRouteConstraint` because I thought it was more viable, mainly because it can resolve mysterious `AmbiguousMatchExceptions` without apparent cause and doesn't add much overhead natively. The fact is that the routing mechanism works very well, is stable, and is optimized to the max, so it might not be feasible to change it just to fix a simple issue. The route precedence mechanism works very well, and I don't think we can consider it a "bug," but rather a limitation. Regarding the Analyzer, if I'm not mistaken, it already has an alert that warns about similar templates, but I don't think it usually goes into much detail, only considering the template literal.

sami-daniel commented:
Okay, this new implementation has a few issues that I've identified so far, but I'm not sure how to proceed. The first one I've been able to report is whether it's prudent/necessary to recreate a new `ConstraintMap` variable identical to the value in `RouteOptions`. The code uses this manually created variable as a fallback in case we can't resolve `RouteOptions` and obtain the `ConstraintMap`, but I'm not sure if that's appropriate.

Another point I'm not sure how to proceed with is how we should proceed with `RegexRouteConstraint`. I've seen that they're handled specifically for some reason, I imagine to avoid browser support issues or some security problem.
